### PR TITLE
Model additions for planning systems/procedures/steps/phases

### PR DESCRIPTION
**Pull request by invisig0th**

### CODE REVIEW

File: synapse/models/planning.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+                    ('outputs', ('array', {'type': 'plan:procedure:variable'}), {
+                        'doc': 'An array of variables defined in this step.'}),
+
+                    ('techniques', ('array', {'type': 'ou:technique'}), {
+                        'doc': 'An array of techniques used when executing this step.'}),
+
+                    ('links', ('array', {'type': 'plan:procedure:link', 'uniq': True}), {
+                        'doc': 'An array of links to subsequent steps.'}),
+
+                    ('procedure', ('plan:procedure', {}), {
```
Comment (infime) [inline synapse/models/planning.py:126]: procedure duped

File: synapse/models/planning.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                        'doc': 'The procedure which includes the step.'}),
+
+                    ('title', ('str', {}), {
+                        'ex': 'Scan the IPv4 address range for open ports'
+                        'doc': 'The title of the step.'}),
+
+                    ('summary', ('str', {}), {
+                        'doc': 'A summary of the tasks executed within the step.'}),
+
+                    ('outputs', ('array', {'type': 'plan:procedure:variable'}), {
```
Comment (infime) [inline synapse/models/planning.py:117]: uniq?

File: synapse/models/planning.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+                        'doc': 'A URL which links to the full documentation about the phase.'}),
+
+                    ('system', ('plan:system', {}), {
+                        'doc': 'The planning system which defines this phase.'}),
+                )),
+                ('plan:procedure:type:taxonomy', {}, {}),
+                ('plan:procedure', {}, (
+
+                    ('title', ('str', {}), {
+                        'ex': 'Network Reconisence Procedure'
```
Comment (infime) [inline synapse/models/planning.py:64]: Reconnaissance

File: synapse/models/planning.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                    ('updated', ('time', {}), {}),
+                        'doc': 'The time the procedure was last updated.'}),
+
+                    ('version', ('it:semver', {}), {
+                        'doc': 'The version of the procedure.'}),
+
+                    ('type', ('plan:procedure:type:taxonomy', {}), {
+                        'doc': 'A type classification for the procedure.'}),
+
+                    ('inputs', ('array', {'type': 'plan:procedure:variable'}), {
```
Comment (infime) [inline synapse/models/planning.py:82]: uniq?

File: synapse/models/planning.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+                    ('links', ('array', {'type': 'plan:procedure:link', 'uniq': True}), {
+                        'doc': 'An array of links to subsequent steps.'}),
+
+                    ('procedure', ('plan:procedure', {}), {
+                        'doc': 'The procedure which defines the step.'}),
+                )),
+                ('plan:procedure:link', {}, (
+
+                    ('condition', ('bool', {}), {
```
[the remark this answers is not in the document]
Comment (invisig0th) [reply synapse/models/planning.py:131]: The boolean value represents the actual condition, not whether or not the link is conditional :)

File: synapse/models/planning.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+                    ('author' ('ps:contact', {}), {
+                        'doc': 'The contact of the authoring person or organization.'}),
+                    ('created', ('time', {}), {
+                        'doc': 'The time the planning system was first created.'}),
+                    ('url', ('inet:url', {}), {
+                        'doc': 'The primary URL which documents the planning system.'}),
+                )),
+                ('plan:phase', {}, (
+                    ('title', ('str', {}), {
+                        'ex': 'Reconisence Phase'
```
Comment (mbvertex) [inline synapse/models/planning.py:45]: typo - reconnaissance

File: synapse/models/planning.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+                        'doc': 'The title of the phase.'}),
+
+                    ('summary', ('str', {}), {
+                        'doc': 'A summary of the definition of the phase.'}),
+
+                    ('index', ('int', {}), {
```
Comment (mbvertex) [inline synapse/models/planning.py:51]: When we use index here, are we using it to note that this is phase 3 out of 4, etc?

Comment (invisig0th) [reply synapse/models/planning.py:51]: Yep! It's there to provide an arbitrary sorting order.

File: synapse/models/planning.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+    def getModelDefs(self):
+        return (('plan', {
+            'types': (
+                ('plan:system', ('guid', {}), {
+                    'doc': 'A planning or behavioral analysis system that defines phases and procedures.'}),
+
+                ('plan:phase', ('guid', {}), {
+                    'doc': 'A phase within a planning system which may be used to group steps.'}),
+
```
[the remark this answers is not in the document]
Comment (invisig0th) [reply synapse/models/planning.py:13]: It's designed to allow you to optionally group the steps within a procedure into phases.

File: synapse/models/planning.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+                    ('version', ('it:semver', {}), {
+                        'doc': 'The version of the planning system.'}),
+
+                    ('url', ('inet:url', {}), {
+                        'doc': 'The primary URL which documents the planning system.'}),
+                )),
+                ('plan:phase', {}, (
+                    ('title', ('str', {}), {
+                        'ex': 'Reconisence Phase',
```
Comment (reign-git) [inline synapse/models/planning.py:55]: Reconnaissance misspelled

File: synapse/models/planning.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+                ('plan:procedure:step', ('guid', {}), {
+                    'doc': 'A step within a procedure.'}),
+
+                ('plan:procedure:link', ('guid', {}), {
+                    'doc': 'A link between steps in a procedure.'}),
+            ),
+
+            'forms': (
+                ('plan:system', {}, (
```
Comment (reign-git) [inline synapse/models/planning.py:32]: Do we want to add a :desc field?

File: synapse/models/planning.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+                    ('type', ('str', {}), {
+                        'doc': 'The type for the input. Types are specific to the planning system.'}),
+
+                    ('default', ('data', {}), {
+                        'doc': 'The optional default value if the procedure is invoked without the input.'}),
+
+                    ('procedure', ('plan:procedure', {}), {
+                        'doc': 'The procedure which defines the variable.'}),
+                )),
+                ('plan:procedure:step', {}, (
```
Comment (reign-git) [inline synapse/models/planning.py:118]: Does this need an index as well to account for ordering?

File: synapse/models/planning.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+                    ('firststep', ('plan:procedure:step', {}), {
+                        'doc': 'The first step in the procedure.'}),
+                )),
+                ('plan:procedure:variable', {}, (
+
+                    ('name', ('str', {}), {
+                        'doc': 'The name of the variable.'}),
+
+                    ('type', ('str', {}), {
```
Comment (Cisphyx) [inline synapse/models/planning.py:109]: Worth making this a taxonomy type?

Comment (vEpiphyte) [reply synapse/models/planning.py:109]: Every time we do type as a `str`, we almost always regret it.

File: synapse/models/planning.py (hunk 10 above)
Comment (vEpiphyte) [reply synapse/models/planning.py:32]: We have 44 properties with the relname "summary" and 97 properties with the relname "desc". We should probably lean into one of them for consistency.

File: synapse/models/planning.py (hunk 12 above)
Comment (invisig0th) [reply synapse/models/planning.py:109]: This is not a classic use of `:type` like taxonomy, this is designed to capture a type from a programatic type system such as `int` and `str` so a taxonomy does not make sense here. 👍

File: synapse/models/planning.py (hunk 11 above)
Comment (invisig0th) [reply synapse/models/planning.py:118]: No, their order is dictated by the decision tree starting at `plan:procedure:firststep` and is not a linear progression that could be represented by a number.

File: synapse/models/planning.py (hunk 9 above)
Comment (invisig0th) [inline synapse/models/planning.py:55]:
```suggestion
                        'ex': 'Reconnaissance Phase',
```

File: synapse/models/planning.py (hunk 10 above)
Comment (invisig0th) [reply synapse/models/planning.py:32]: I could see the utility of adding a `:summary` field here. I'll tweak it. 👍 @vEpiphyte i think the use of `summary` has been growing for specific use cases that are a bit more abstract or self defined. I've also been veering toward property names that are a tiny bit friendlier vs saving a couple characters ( partially due to indexing pressure relief via the use of abrv translations everywhere :D )

File: synapse/models/planning.py (hunk 9 above)
Comment (invisig0th) [reply synapse/models/planning.py]: lol, thanks :)

File: synapse/models/planning.py (hunk 12 above)
Comment (vEpiphyte) [reply synapse/models/planning.py:109]: Do we want to `onespace` and `lower` the value then? Avoid `Int` vs `int` vs ` int ` problems.

**Review by MichaelSquires — APPROVED** [review]

**Review by Cisphyx — APPROVED** [review]

**Pull request merged**